packages/react-aria-components/src/Button.tsx

```diff
@@ -77,31 +112,116 @@ export const ButtonContext = createContext<ContextValue<ButtonContextValue, HTML
 
 function Button(props: ButtonProps, ref: ForwardedRef<HTMLButtonElement>) {
   [props, ref] = useContextProps(props, ref, ButtonContext);
+  props = disablePendingProps(props);
   let ctx = props as ButtonContextValue;
   let {buttonProps, isPressed} = useButton(props, ref);
   let {focusProps, isFocused, isFocusVisible} = useFocusRing(props);
   let {hoverProps, isHovered} = useHover(props);
+  let [isPendingVisible, setIsPendingVisible] = useState(false);
   let renderProps = useRenderProps({
     ...props,
-    values: {isHovered, isPressed, isFocused, isFocusVisible, isDisabled: props.isDisabled || false},
+    values: {isHovered, isPressed, isFocused, isFocusVisible, isDisabled: props.isDisabled || false, isPendingVisible},
     defaultClassName: 'react-aria-Button'
   });
 
+  let formatter = useLocalizedStringFormatter(intlMessages, 'react-aria-components');
+  let {isPending, pendingDelay = 1000, renderPendingState} = ctx;
+  // an aria label will block children and their labels from being read, this is undesirable for pending state
+  let hasAriaLabel = !!buttonProps['aria-label'] || !!buttonProps['aria-labelledby'];
+  let spinnerId = useId();
+  let buttonId = useId(buttonProps.id);
+  let containerId = useId();
+
+  useEffect(() => {
+    let timeout: ReturnType<typeof setTimeout>;
+
+    if (isPending) {
+      // Start timer when isPending is set to true.
+      timeout = setTimeout(() => {
+        setIsPendingVisible(true);
+      }, pendingDelay);
+    } else {
+      // Exit loading state when isPending is set to false. */
+      setIsPendingVisible(false);
+    }
+    return () => {
+      // Clean up on unmount or when user removes isPending prop before entering loading state.
+      clearTimeout(timeout);
+    };
+  }, [isPending]);
```

Comment:
interesting, how would they handle this with a CSS animation? in general, nothing happens during this timeout. It's the space of time between showing the normal button and showing the pending state. What transition would we put in there?

Reply:
You could animate `visibility` for example, or `opacity`

Reply:
Ah, interesting, hadn't thought about visibility as a candidate. Since it doesn't actually animate, just flips at the end of the animation.
Ok, I can give that a try

Reply:
how would we know which animation signaled pending?
if for example, a ripple animation was used, that would interfere no? I don't think we want to count of a specifically named animation

Reply:
also, one of the things we need is for visibility to happen a frame after isPending turns true due to some hacking around AT

Reply:
Wouldn't the animation go on the spinner itself?

```css
@keyframes load {
  from {
    visibility: hidden;
  }

  to {
    visibility: visible;
  }
}

.spinner {
  animation: load 1s;
}
```

With something like that, the spinner should appear 1s after it mounts. And due to visibility: hidden, it wouldn't be in the accessibility tree until then right?

Reply:
but we also have a spinning animation on the spinner? and maybe there's a fade in after the switch from hidden to visible?

Reply:
You can have multiple animations on the same element, you just comma separate them

Reply:
Right, how do I know which one I'm listening for with onAnimationEnd? They come through as names, not properties

Reply:
Why do you need to know when it ends?

Reply:
I was thinking to use it as the timer to trigger all the state updates, but you were suggesting moving everything up to be instant.